### PR TITLE
Add pin pad type states to Sercom module

### PR DESCRIPTION
**Pull request by keithduncan**
I thought I’d open this sooner rather than later to see if there’s any interest in merging it upstream.

I’ve tried to add type states to the sercom module to allow expressing board connections at the type level. I haven’t used rust macros previously so there’s probably a neater way to accomplish this than what I’ve done. I’ve also only changed the `samd21` code variants so far as that’s what I’m using, if this is something you'd be interested in I can add it to the `samd51` variant modules too.

I’m using the Arduino MKR WiFi 1010 and Arduino MKR NB 1500 for a project and wanted to codify how the on board ECC508, NINA W10 and SARA R410M are connected to the MCU. Similar to how the GPIO module restricts state transitions using types I wanted to export driver types from my board support crates that would restrict the driver to being constructed with the correct sercom mode using the correct physical pins in the right function.

Before a driver would be exported as:

```
pub extern crate ecc508_driver as ecc508;
pub type Ecc508 = ecc508::Ecc508<mcu::sercom::I2CMaster0>;
```

Now this can be exported as:

```
pub extern crate ecc508_driver as ecc508;
pub type Ecc508 = ecc508::Ecc508<
    mcu::sercom::I2CMaster0<
        mcu::sercom::Sercom0Pad0<mcu::gpio::Pa8<mcu::gpio::PfC>>,
        mcu::sercom::Sercom0Pad1<mcu::gpio::Pa9<mcu::gpio::PfC>>
    >
>;
```

Let me know if this is something you’re interested in and if there are any changes you’d like to see.

---

Todo:

- [x] Document the macros and explain what they generate
- [x] Ensure CI is covering the samd51 hal variants
- [x] Add type states to samd51 module hal variants
- [ ] Review SPI pad configuration and DIPO/DOPO values closely 👀 
- [ ] Review UART pad configuration and RXPO/TXPO values closely 👀

### CODE REVIEW

**Comment by sajattack** [comment]
Looks like the SPIMaster# and I2CMaster#s in the BSPs need to be updated, but otherwise, looks good. The more typestates the better.

File: hal/src/sercom/uart.rs [hunk 1]
```diff
@@ -11,242 +11,234 @@ use hal::blocking::serial::{write::Default, Write};
...
+                pub cts: CTS,
+            }
+        }
 
-            &$Type::Rx3Tx0{..} => (3, 0),
-            &$Type::Rx3Tx2{..} => (3, 1),
+        macro_rules! padout {
+            ($rxpo_txpo:expr => $pad0:ident, $pad1:ident) => {
+                $crate::paste::item! {
+                    impl<PIN0, PIN1> From<([<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>)> for [<$Type Padout>]<[<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>, (), ()> {
```
Comment (wez) [inline hal/src/sercom/uart.rs]: these impls could also do with a brief comment to help future grok-ability

File: hal/src/sercom/spi.rs [hunk 2]
```diff
@@ -5,249 +5,203 @@ use crate::target_device::{PM, SERCOM0, SERCOM1, SERCOM2, SERCOM3};
...
+            }
         }
 
-        Self {
-            pinout,
-            sercom,
+        macro_rules! padout {
+            ($dipo_dopo:expr => $pad0:ident, $pad1:ident, $pad2:ident) => {
+                $crate::paste::item! {
+                    impl<PIN0, PIN1, PIN2> From<([<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>, [<$Sercom $pad2>]<PIN2>)> for [<$Type Padout>]<[<$Sercom $pad0>]<PIN0>, [<$Sercom $pad1>]<PIN1>, [<$Sercom $pad2>]<PIN2>> {
```
[the remark this answers is not in the document]
Comment (keithduncan) [reply hal/src/sercom/spi.rs]: :+1: it is extremely dense 😅

File: hal/src/sercom/spi.rs [hunk 3]
```diff
@@ -5,249 +5,203 @@ use crate::target_device::{PM, SERCOM0, SERCOM1, SERCOM2, SERCOM3};
...
-}
-
-impl FullDuplex<u8> for $Type {
-    type Error = Error;
 
-    fn read(&mut self) -> nb::Result<u8, Error> {
-        let status = self.spi().status.read();
-        if status.bufovf().bit_is_set() {
-            return Err(nb::Error::Other(Error::Overrun));
+        padout!((0, 1) => Pad0, Pad2, Pad3);
```
[the remark this answers is not in the document]
Comment (keithduncan) [reply hal/src/sercom/spi.rs]: :+1: I’ve found the generated docs helpful to check that the macros implement what you expect too. I’ve added a todo ☝️ to have these and the UART ones reviewed closely.

File: hal/src/sercom/uart.rs (hunk 1 above)
Comment (keithduncan) [reply hal/src/sercom/uart.rs]: :+1:

**Comment by keithduncan** [comment]
I think this is ready to review again.

I’ve added docs for the macros and the new Sercom mode pad configuration types, CI is passing in all the variants except `nightly` on travis which is broken for thumbv6 for the time being. (An updated nightly has fixed this already but I don’t know how often Travis rerolls their images).

Still to do is getting some more :eyes: on the DIPO/DOPO and RXPO/TXPO values and their corresponding Pad configurations.

**Comment by sajattack** [comment]
@wez do you have time to look at this? You'd know the DIPO/DOPO better than I would.

**Pull request merged**